doc/dev/tmem.md

```diff
@@ -0,0 +1,1167 @@
+
+<details>
+<summary>32x32b:</summary>
+
+![32x32b](https://docs.nvidia.com/cuda/parallel-thread-execution/_images/tcgen05-mma-fragment-3232b.png)
+</details>
+
+<details>
+<summary>16x64b:</summary>
+
+![16x64b](https://docs.nvidia.com/cuda/parallel-thread-execution/_images/tcgen05-mma-fragment-1664b.png)
+</details>
+
+<details>
+<summary>16x128b:</summary>
+
+![16x128b](https://docs.nvidia.com/cuda/parallel-thread-execution/_images/tcgen05-mma-fragment-16128b.png)
+</details>
+
+
+<details>
+<summary>16x256b:</summary>
+
+![16x256b](https://docs.nvidia.com/cuda/parallel-thread-execution/_images/tcgen05-mma-fragment-16256b.png)
+</details>
+
+
+<details>
+<summary>16x32bx2:</summary>
+
+![16x32bx2](https://docs.nvidia.com/cuda/parallel-thread-execution/_images/tcgen05-mma-fragment-1632b2.png)
+</details>
+
+Besides threads in a warp must satisfy specific pattern, another restriction of
+the TMem<->register transfer is: not all warps can access all lanes of the tensor
+memory. The entire 128 lanes of the tensor memory is divided into 4
+subpartitions, each has 32 lanes. The warp `i` can only access the subpartition
+`i % 4`.
+
+With the above restrictions in mind, let's take a look at a few examples of how
+NOT to schedule TMem load and store:<!-- */ //-->\
+```cpp
+TEST_F(TMemTutorialC, NotWarpCollective) {
+  NOT_IMPLEMENTED
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeContigConcreteTensor({16, 2});
+  fusion.addInput(tv0);
+  auto tv1 = set(tv0);
+  auto tv2 = set(tv1);
+  auto tv3 = set(tv2);
+  auto tv4 = set(tv3);
+  fusion.addOutput(tv4);
+  tv2->setMemoryType(MemoryType::Tensor);
+
+  tv4->axis(0)->parallelize(ParallelType::TIDx);
+  scheduler_utils::parallelizeAllLike(tv4);
+
+  tv2->setTMemDimSepPos(1);
+
+  EXPECT_THAT(
+      [&]() { KernelExecutor().compile(&fusion); },
+      ::testing::ThrowsMessage<std::runtime_error>(
+          ::testing::HasSubstr("TMem load/store must be warp collective.")));
+} /*
+```
+
+The above example is invalid because there are only 16 threads in the kernel.
+Warp collective operations require at least a whole warp to run.<!-- */ //-->\
+```cpp
+TEST_F(TMemTutorialC, NotContiguous) {
+  NOT_IMPLEMENTED
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeContigConcreteTensor({64, 2, 2});
+  fusion.addInput(tv0);
+  auto tv1 = set(tv0);
+  auto tv2 = set(tv1);
+  auto tv3 = set(tv2);
+  auto tv4 = set(tv3);
+  fusion.addOutput(tv4);
+  tv2->setMemoryType(MemoryType::Tensor);
+
+  tv4->axis(0)->parallelize(ParallelType::TIDx);
+  tv4->axis(1)->parallelize(ParallelType::TIDy);
+  scheduler_utils::parallelizeAllLike(tv4);
+
+  tv2->setTMemDimSepPos(2);
+
+  EXPECT_THAT(
+      [&]() { KernelExecutor().compile(&fusion); },
+      ::testing::ThrowsMessage<std::runtime_error>(::testing::HasSubstr(
+          "Invalid data access pattern in TMem load/store.")));
+} /*
+```
+
+The above example is invalid because the tensor memory is not accessed in any of
```

Comment:
Why this case is stride-2?

Reply:
Updated the doc to explain this:

because the lane allocation domain is `[TIDx{64}, TIDy{2}]`, where `TIDx` is not the innermost

Reply:
That makes sense.